silver/models/documents/base.py

```diff
@@ -493,3 +527,99 @@ def total_before_tax_in_transaction_currency(self):
+        Transaction = get_model('silver.Transaction')
+
+        return self.total_in_transaction_currency - sum([
+            transaction.amount
+            for transaction in self.transactions.filter(state__in=[
+                Transaction.States.Initial,
+                Transaction.States.Pending,
+                Transaction.States.Settled
+            ])
+        ])
+
+
+def create_transaction_for_document(document):
+    # get a usable, recurring payment_method for the customer
+    PaymentMethod = get_model('silver.PaymentMethod')
+    Transaction = get_model('silver.Transaction')
+
+    payment_methods = PaymentMethod.objects.filter(
+        canceled=False,
+        verified=True,
+        customer=document.customer
+    )
+    for payment_method in payment_methods:
+        try:
+            return Transaction.objects.create(document=document,
+                                              payment_method=payment_method)
+        except ValidationError:
+            continue
+
+
+@receiver(post_transition)
+def post_transition_callback(sender, instance, name, source, target, **kwargs):
+    if not isinstance(instance, BillingDocumentBase):
+        return
+
+    document = instance
+    setattr(document, '.recently_transitioned', target)
+
+    document.save()
+
+
+@receiver(post_save)
+def post_document_save(sender, instance, created=False, **kwargs):
+    if not isinstance(instance, BillingDocumentBase):
+        return
+
+    document = instance
+
+    if hasattr(document, '.recently_transitioned'):
+        delattr(document, '.recently_transitioned')
+
+        # Transition related document too, if needed
+        if document.related_document and document.state != document.related_document.state:
+            state_transition_map = {
+                BillingDocumentBase.STATES.ISSUED: 'issue',
+                BillingDocumentBase.STATES.CANCELED: 'cancel',
+                BillingDocumentBase.STATES.PAID: 'pay'
+            }
+            transition_name = state_transition_map[document.state]
+
+            bound_transition_method = getattr(document.related_document, transition_name)
+            bound_transition_method()
+
+        # Create a transaction if the document was recently issued
+        if (document.state == BillingDocumentBase.STATES.ISSUED and
+                settings.SILVER_AUTOMATICALLY_CREATE_TRANSACTIONS):
+            # But only if there is no pending transaction
+            Transaction = get_model('silver', 'Transaction')
+
+            # The related document might have the only reference to an existing transaction
+            if not (document.related_document or document).transactions.filter(
+                state__in=[Transaction.States.Pending,
+                           Transaction.States.Initial,
+                           Transaction.States.Settled]
+            ):
+                create_transaction_for_document(document)
+
+        # Generate a PDF
+        document.set_pdf_for_generation()
```

Comment:
:+1:
I would name it `mark_for_generation` (in this way you will leave some options to have more renders: pdf, html, xml. Maybe, in a near feature, each invoice will have a well know format, xml based, making them easier to import/export (I think that the Czech Republic has this system implemented)).